frontend/src/utilities/__tests__/clusterUtils.spec.ts

```diff
@@ -0,0 +1,117 @@
+import * as clusterInfo from '~/redux/selectors/clusterInfo';
+import { getOpenShiftConsoleServerURL, useOpenShiftURL } from '~/utilities/clusterUtils';
+import { testHook } from '~/__tests__/unit/testUtils/hooks';
+let originalLocation: Location;
+
+const setupWindowLocation = (hostname: string, protocol: string, port: string) => {
+  Object.defineProperty(window, 'location', {
+    writable: true,
+    value: { ...originalLocation, hostname, protocol, port },
+  });
+};
+
+jest.mock('~/utilities/const', () => ({
+  get DEV_MODE() {
+    return false;
+  },
+}));
+
+jest.mock('~/redux/selectors/clusterInfo', () => ({
+  ...jest.requireActual('~/redux/selectors/clusterInfo'), // Use the actual module for other exports
+  useClusterInfo: jest.fn(),
+}));
+
+const devModeMock = jest.spyOn(jest.requireMock('~/utilities/const'), 'DEV_MODE', 'get');
+
+beforeAll(() => {
+  originalLocation = { ...window.location };
+});
+
+afterAll(() => {
+  Object.defineProperty(window, 'location', {
+    writable: true,
+    value: originalLocation,
+  });
+});
+
+describe('getOpenShiftConsoleServerURL', () => {
+  beforeEach(() => {
+    setupWindowLocation('localhost', 'https:', '443');
+  });
+
+  afterEach(() => {
+    // Restore window.location to its original state
+    Object.defineProperty(window, 'location', {
+      writable: true,
+      value: originalLocation,
+    });
+  });
+
+  it('should construct URL based on window location when no apiURL is provided', () => {
+    setupWindowLocation('api.example.com', 'https:', '443');
+    const result = getOpenShiftConsoleServerURL();
+
+    // Assert the constructed URL
+    expect(result).toBe('https://console-openshift-console.example.com:443');
+  });
+
+  it('should construct URL with DEV_MODE true and provided apiURL', () => {
+    devModeMock.mockReturnValue(true);
+    const apiURL = 'https://api.example.com:8443';
+    const result = getOpenShiftConsoleServerURL(apiURL);
+
+    // Assert the constructed URL
+    expect(result).toBe('https://console-openshift-console.apps.example.com');
+  });
+
+  it('should return null when hostParts.length is less than 2', () => {
+    const result = getOpenShiftConsoleServerURL();
+    // Assert
+    expect(result).toBeNull();
+  });
+});
+
+describe('useOpenShiftURL', () => {
+  beforeEach(() => {
+    setupWindowLocation('localhost', 'https:', '443');
+  });
+
+  afterEach(() => {
+    // Restore window.location to its original state
+    Object.defineProperty(window, 'location', {
+      writable: true,
+      value: originalLocation,
+    });
+  });
+  it('should return the constructed URL from useClusterInfo', () => {
+    const serverURL = 'https://api.example.com';
+
+    // Manually mock the useClusterInfo function
+    const mockUseClusterInfo = jest.spyOn(clusterInfo, 'useClusterInfo');
```

Comment:
Using multiple methods of mocking when one will do. No need to use `spyOn` here.